### ISSUE TITLE
Develop deeplearning models and hyperparameter search for SKAB anomaly prediction(early warning)

### ISSUE THREAD

**Issue by XuLi111111** [XuLi111111/predictive_maintenance_anomaly_detection#27]
Main tasks:

1. Model development
- Design a deep learning model for time-series anomaly prediction
- Support binary classification based on future anomaly occurrence

2. Training pipeline
- Use the processed SKAB dataset with sliding window inputs
- Maintain a consistent data preprocessing and training workflow

3. Model selection
- Explore different model configurations
- Select a suitable model for the task

4. Evaluation
- Evaluate model performance on validation and test sets
- Compare results across different configurations

**Comment by XuLi111111** [comment]
<img width="1181" height="759" alt="Image" src="https://github.com/user-attachments/assets/cf797a5e-7194-4971-9a4b-f8c86afb80a1" />

Update on this issue:

1. Model design
I adopted a Transformer-based architecture to capture temporal dependencies in time-series data. 
In addition, I introduced hand-crafted statistical features and fused them with the learned sequence representations.

The hand-crafted features include:
- mean, standard deviation
- max and min values
- first and last timestep values
- difference between last and first values
- temporal difference statistics (mean and std of consecutive differences)
- absolute maximum values

These features are designed to capture both global trends and local variations within each time window.

During development, I also explored alternative designs. For example, a BiLSTM-only model (without Transformer encoder blocks) was tested, but showed weaker performance. Based on both empirical results and insights from relevant literature, I adopted the feature fusion approach as the final design.

2. Ablation analysis
I conducted ablation experiments to evaluate the contribution of hand-crafted features. 
The results show that incorporating these features improves anomaly detection performance, increasing anomaly recall by approximately 1.7 percentage points.

3. Hyperparameter search
After finalising the model structure, I performed a grid search to identify suitable hyperparameter configurations.

The search space included:
- d_model ∈ {32, 48, 64}
- nhead ∈ {4, 8}
- feed-forward dimension ∈ {64, 96, 128, 160}
- dropout ∈ {0.03, 0.08, 0.12, 0.18}
- number of layers = 2

In total, 50+ model configurations were evaluated.

The best-performing model was:
TransformerFusionLite_L2_D32_H8_FF128_DO003  
Best threshold (validation): 0.68

4. Final evaluation
The final model achieves the following performance on the test set:

- Precision (anomaly): 0.9890  
- Recall (anomaly): 0.8706  
- F1-score (anomaly): 0.9260  
- Weighted F1: 0.9478  
- Accuracy: 0.9485  
- False Positive Rate: 0.0057  
- False Negative Rate: 0.1294  

Confusion Matrix:
[[2628, 15],
 [201, 1352]]

5. Outcome
The model achieves strong and well-balanced performance across all evaluation metrics, particularly maintaining high anomaly recall while keeping a low false positive rate.

The model has been validated and confirmed as suitable for downstream integration (e.g., deployment in the web application).

Model structure diagram is attached.
